webcrypto/aes.go

```diff
@@ -0,0 +1,155 @@
+package webcrypto
+
+import (
+	"crypto/rand"
+	"fmt"
+	"strings"
+
+	"github.com/dop251/goja"
+)
+
+// AesKeyAlgorithm represents the AES algorithm.
+type AesKeyAlgorithm struct {
+	KeyAlgorithm
+
+	// The length, in bits, of the key.
+	Length int `json:"length"`
+}
+
+// AesKeyGenParams represents the object that should be passed as
+// the algorithm parameter into `SubtleCrypto.generateKey`, when generating
+// an AES key: that is, when the algorithm is identified as any
+// of AES-CBC, AES-CTR, AES-GCM, or AES-KW.
+type AesKeyGenParams struct {
+	Algorithm
+
+	// Length holds (a Number) the length of the key, in bits.
+	Length int `json:"length"`
+}
+
+// Ensure AesKeyGenParams implements the From interface.
+var _ From[map[string]interface{}, AesKeyGenParams] = AesKeyGenParams{}
+
+// From fills the AesKeyGenParams from the given dictionary object.
+func (a AesKeyGenParams) From(dict map[string]interface{}) (AesKeyGenParams, error) {
+	nameFound := false
+	lengthFound := false
+
+	for key, value := range dict {
+		if strings.EqualFold(key, "name") {
+			name, ok := value.(string)
+			if !ok {
+				return AesKeyGenParams{}, NewError(0, SyntaxError, "name property should hold a string")
+			}
+
+			name = strings.ToUpper(name)
+
+			if !IsAlgorithm(name) {
+				return AesKeyGenParams{}, NewError(0, NotSupportedError, fmt.Sprintf("algorithm %s is not supported", name))
+			}
+
+			a.Name = name
+			nameFound = true
+			continue
+		}
+
+		if strings.EqualFold(key, "length") {
+			length, ok := value.(int64)
+			if !ok {
+				return AesKeyGenParams{}, NewError(0, SyntaxError, "length property should hold a number")
+			}
+
+			a.Length = int(length)
+			lengthFound = true
+			continue
+		}
+	}
+
+	if !nameFound {
+		return AesKeyGenParams{}, NewError(0, SyntaxError, "name property is required")
+	}
+
+	if !lengthFound {
+		return AesKeyGenParams{}, NewError(0, SyntaxError, "length property is required")
+	}
+
+	return a, nil
+}
+
+// Ensure AesKeyGenParams implements the From interface.
+var _ KeyGenerator = &AesKeyGenParams{}
+
+// GenerateKey generates a new AES key.
+func (a *AesKeyGenParams) GenerateKey(
+	rt *goja.Runtime,
+	extractable bool,
+	keyUsages []CryptoKeyUsage,
+) (goja.Value, error) {
+	var (
+		isCBC = a.Algorithm.Name == AESCbc
+		isCTR = a.Algorithm.Name == AESCtr
+		isGCM = a.Algorithm.Name == AESGcm
+		isKW  = a.Algorithm.Name == AESKw
+	)
+	if !isCBC && !isCTR && !isGCM && !isKW {
```

Comment:
This validation should be in the constructor method, not the `GenerateKey`. Otherwise, we're breaking concerns :cry: 

So the constructor should produce a valid object (`AesKeyGenParams`), and later any operation on it should be executed safely without additional checks.

Reply:
You're right 👍🏻